### PR TITLE
Added a check and recomputation of navmesh for non-standard agents

### PR DESCRIPTION
**Pull request by mathfac**
## Motivation and Context
Added a check and recomputation of navmesh for non standard agent size. That will add the support of navigation metrics and shortest path calculation for custom agent configuration.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by 1heart** [comment]
# [Codecov](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466?src=pr&el=h1) Report
> Merging [#466](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466?src=pr&el=desc) into [master](https://codecov.io/gh/facebookresearch/habitat-sim/commit/74d85018df989ec3c35acd76b15e7a5a04899a91&el=desc) will **decrease** coverage by `0.22%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master     #466      +/-   ##
==========================================
- Coverage   58.55%   58.32%   -0.23%     
==========================================
  Files         165      161       -4     
  Lines        7327     7163     -164     
  Branches       84       84              
==========================================
- Hits         4290     4178     -112     
+ Misses       3037     2985      -52     
```

| Flag | Coverage Δ | |
|---|---|---|
| #CPP | `53.09% <ø> (-0.54%)` | :arrow_down: |
| #JavaScript | `10.00% <ø> (ø)` | :arrow_up: |
| #Python | `79.40% <33.33%> (-0.18%)` | :arrow_down: |

| [Impacted Files](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [src/esp/bindings/SceneBindings.cpp](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466/diff?src=pr&el=tree#diff-c3JjL2VzcC9iaW5kaW5ncy9TY2VuZUJpbmRpbmdzLmNwcA==) | | |
| [src/esp/bindings/SensorBindings.cpp](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466/diff?src=pr&el=tree#diff-c3JjL2VzcC9iaW5kaW5ncy9TZW5zb3JCaW5kaW5ncy5jcHA=) | | |
| [src/esp/sensor/RedwoodNoiseModel.h](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466/diff?src=pr&el=tree#diff-c3JjL2VzcC9zZW5zb3IvUmVkd29vZE5vaXNlTW9kZWwuaA==) | | |
| [src/esp/bindings/OpaqueTypes.h](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466/diff?src=pr&el=tree#diff-c3JjL2VzcC9iaW5kaW5ncy9PcGFxdWVUeXBlcy5o) | | |
| [src/esp/sensor/VisualSensor.h](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466/diff?src=pr&el=tree#diff-c3JjL2VzcC9zZW5zb3IvVmlzdWFsU2Vuc29yLmg=) | `20.00% <0.00%> (+8.23%)` | :arrow_up: |
| [src/esp/sensor/Sensor.h](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466/diff?src=pr&el=tree#diff-c3JjL2VzcC9zZW5zb3IvU2Vuc29yLmg=) | `81.81% <0.00%> (+12.58%)` | :arrow_up: |
| [src/esp/scene/SemanticScene.h](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466/diff?src=pr&el=tree#diff-c3JjL2VzcC9zY2VuZS9TZW1hbnRpY1NjZW5lLmg=) | `30.76% <0.00%> (+14.44%)` | :arrow_up: |
| [src/esp/scene/ObjectControls.h](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466/diff?src=pr&el=tree#diff-c3JjL2VzcC9zY2VuZS9PYmplY3RDb250cm9scy5o) | `50.00% <0.00%> (+16.66%)` | :arrow_up: |
| [src/esp/scene/SceneNode.h](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466/diff?src=pr&el=tree#diff-c3JjL2VzcC9zY2VuZS9TY2VuZU5vZGUuaA==) | `100.00% <0.00%> (+22.22%)` | :arrow_up: |
| [src/esp/scene/SceneGraph.h](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466/diff?src=pr&el=tree#diff-c3JjL2VzcC9zY2VuZS9TY2VuZUdyYXBoLmg=) | `100.00% <0.00%> (+40.00%)` | :arrow_up: |
| ... and [1 more](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466/diff?src=pr&el=tree-more) | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466?src=pr&el=footer). Last update [74d8501...f25fdc2](https://codecov.io/gh/facebookresearch/habitat-sim/pull/466?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: habitat_sim/simulator.py [hunk 1]
```diff
@@ -145,6 +145,21 @@ def _config_pathfinder(self, config: Configuration):
...
+        ) or not np.isclose(agent_legacy_config.height, default_agent_config.height):
+            logger.info(
+                f"Recomputing navmesh for agent's height {default_agent_config.height} and radius"
+                f" {default_agent_config.radius}."
+            )
+            navmesh_settings = hsim.NavMeshSettings()
+            navmesh_settings.set_defaults()
+            navmesh_settings.agent_radius = default_agent_config.radius
+            navmesh_settings.agent_height = default_agent_config.height
+            self._sim.recompute_navmesh(self._sim.pathfinder, navmesh_settings)
```
Comment (erikwijmans) [inline habitat_sim/simulator.py]:
```suggestion
            self.recompute_navmesh(self.pathfinder, navmesh_settings)
```

**Comment by mathfac** [comment]
Added the test scenario to guard the functionality.

File: habitat_sim/simulator.py [hunk 2]
```diff
@@ -145,6 +145,21 @@ def _config_pathfinder(self, config: Configuration):
                 f"Could not find navmesh {navmesh_filenname}, no collision checking will be done"
             )
 
+        agent_legacy_config = habitat_sim.AgentConfiguration()
```
Comment (erikwijmans) [inline habitat_sim/simulator.py]: I'm confused how this works.  We don't have an `import habitat_sim` in this file, so that should lead to an error?

Comment (mathfac) [reply habitat_sim/simulator.py]: Switched from using `habitat_sim.*`.

File: habitat_sim/simulator.py [hunk 3]
```diff
@@ -145,6 +145,21 @@ def _config_pathfinder(self, config: Configuration):
...
+        ) or not np.isclose(agent_legacy_config.height, default_agent_config.height):
+            logger.info(
+                f"Recomputing navmesh for agent's height {default_agent_config.height} and radius"
+                f" {default_agent_config.radius}."
+            )
+            navmesh_settings = NavMeshSettings()
+            navmesh_settings.set_defaults()
+            navmesh_settings.agent_radius = default_agent_config.radius
+            navmesh_settings.agent_height = default_agent_config.height
+            self.recompute_navmesh(self.pathfinder, navmesh_settings)
```
Comment (Skylion007) [inline habitat_sim/simulator.py]: Does recompute_navmesh work with blind agents (when the mesh isn't loaded into memory)? I have encountered some errors in the past under that scenario. Should we throw a more informative error for blind agents or throw a warning and use the incorrect navmesh?

**Comment by erikwijmans** [comment]
@aclegg3 added this message here: https://github.com/facebookresearch/habitat-sim/blob/master/src/esp/sim/Simulator.cpp#L374

**Comment by Skylion007** [comment]
As implemented, if I wanted to pre-generate a NavMesh for a blind agent of non-standard size and use that, this would force a nav-mesh recomputation and an assertion failure right? 

Perhaps we should add a check if the agent is blind into the conditional for whether to call recompute_namesh? This would allow the user to specify an appropriate custom nav mesh for a blind agent if needed.

**Comment by mathfac** [comment]
This PR isn't tackling improvement or settings of the blind agents. We have assert to notify users that navmesh recompute isn't supported. We should consider to disable optimization for Blind agents or make it configurable as separate PR cc @erikwijmans.

**Pull request merged**